packages/i18n/src/index.tsx

```diff
@@ -157,10 +158,98 @@ export const createI18nContext = (
      */
     dict: (lang: string) => deepReadObject(dict, lang)
   };
-  return [translate, actions];
+  return [translate, actions as any];
 };
 
 type I18nContextInterface = ReturnType<typeof createI18nContext>;
 
 export const I18nContext = createContext<I18nContextInterface>({} as I18nContextInterface);
 export const useI18n = () => useContext(I18nContext);
+
+// -------------------------- Chained I18n -------------------
+type I18nFormatArgs = Record<string, string | number>;
+
+export type I18nPath<T> = {
+  [K in keyof T]: T[K] extends Record<string, unknown>
+    ? I18nPath<T[K]>
+    : T[K] extends (...args: any) => string
+    ? (...args: Parameters<T[K]>) => string
+    : T[K] extends string
+    ? (args?: I18nFormatArgs) => string
+    : never;
+};
+
+const buildI18nChain = <T,>(obj: T): I18nPath<T> => {
+  const keys = Object.keys(obj as any) as (keyof T)[];
+  const paths = keys.reduce((acc, key) => {
+    const value = obj[key];
+    if (typeof value === "object") {
+      return {
+        ...acc,
+        [key]: buildI18nChain(value)
+      } as any;
+    } else if (typeof value === "function") {
+      return {
+        ...acc,
+        [key]: obj[key]
+      };
+    }
+    if (typeof value === "string") {
+      return {
+        ...acc,
+        [key]: (args: I18nFormatArgs) => {
+          return value.replace(/{{(.*?)}}/g, (_, key) => deepReadObject(args, key, ""));
+        }
+      };
+    } else {
+      throw new TypeError("Unsupported data format on the keys");
+    }
+  }, {} as Partial<I18nPath<T>>);
+
+  return paths as I18nPath<T>;
+};
+
+export const makeChainedI18nContext = <T extends object, K extends keyof T>(props: {
```

Comment:
Shouldn't there be a required interface for the translations (or directories) object? Instead of `object`. You are making assumptions that all functions will return `strings` and uses `I18nFormatArgs` for parameters so it would be useful to guide users to that.

Reply:
I would love to make this an improved interface. I'm not sure if my typescript skills are good enough to enforce this on the object that is being passed in, but I'll play with it some more. 

The current typescript will resolve any key that doesn't fit the type def into never if you try to use it and there is a type error that is thrown with some helper information.

Reply:
Updated the typing on this and it will now have issues if there is an end key path does not return a string or a function that does not return a string.